### ISSUE TITLE
Job is executed before "depends_on" job is finished when using Job.create and q.enqueue_job

### ISSUE THREAD

**Issue by jtfidje** [rq/rq#1404]
Consider the following snippet below and two running workers:
```python
r = Redis(
    host="localhost",
    port=6379,
    db=0
)

q = Queue(connection=r)

id_1 = str(uuid4())

q.enqueue(func_1, job_id=id_1)
q.enqueue(func_2, depends_on=id_1)
```
In that snippet, everything works as expected; both jobs are enqueued ( one on each worker ) but func_2 does not run before func_1 is finished. 

Now consider the following snippet, still with two workers:
```python
r = Redis(
    host="localhost",
    port=6379,
    db=0
)

q = Queue(connection=r)

id_1 = str(uuid4())
id_2 = str(uuid4())

job_1 = Job.create(func_1, id=id_1, connection=r) 
job_2 = Job.create(func_2, id=id_2, depends_on=id_1, connection=r)
q.enqueue_job(job_1)
q.enqueue_job(job_2)
```
When I run that code both jobs are started immediately, despite setting the `depends_on` argument. 

<br>
<br>

#### Mini bonus question
Why is it necessary to set `connection=` on the jobs when I use `Job.create`? Doesn't the `Queue` instance know what Redis connection to use already? Code fails without me setting `connection` on all the jobs as well.

**Comment by selwin** [comment]
[This PR](https://github.com/rq/rq/pull/1397) adds multi dependency feature to RQ proper. You should try this instead.

**Comment by jtfidje** [comment]
Thanks for answering!

I've looked through the PR, but I cant see how those changes would fix this problem? The only way I managed to fix this was to create a custom Queue class and implement the following function as a replacement for the initial `enqueue_job` call:

```python
def enqueue_custom(
        self, job: CustomJob, pipeline=None, at_front: bool = False
    ) -> CustomJob:

        # If a _dependent_ job depends on any unfinished job, register all the
        # _dependent_ job's dependencies instead of enqueueing it.
        #
        # `Job#fetch_dependencies` sets WATCH on all dependencies. If
        # WatchError is raised in the when the pipeline is executed, that means
        # something else has modified either the set of dependencies or the
        # status of one of them. In this case, we simply retry.
        job.set_status(FlowStatus.QUEUED)

        if job._dependency_id:
            with self.connection.pipeline() as pipe:
                while True:
                    try:

                        pipe.watch(job.dependencies_key)

                        dependencies = job.fetch_dependencies(watch=True, pipeline=pipe)

                        pipe.multi()

                        for dependency in dependencies:
                            if (
                                dependency.get_status(refresh=False)
                                != FlowStatus.FINISHED
                            ):
                                job.set_status(FlowStatus.DEFERRED, pipeline=pipe)
                                job.register_dependency(pipeline=pipe)
                                job.save(pipeline=pipe)
                                job.cleanup(ttl=job.ttl, pipeline=pipe)
                                pipe.execute()
                                return job

                        break
                    except WatchError:
                        continue
        return super().enqueue_job(job, pipeline, at_front=at_front)
```

You have this if-block inside `enqueue_call` as well, and I couldn't get dependencies to work without it.

**Comment by selwin** [comment]
With the PR merged, you can do `queue.enqueue(my_func, depends_on=[job_1, job_2])`

**Comment by jtfidje** [comment]
Yes I understand ( and really like this feature! )
But, in the [docs](https://python-rq.org/docs/jobs/) you give this example:

```python
from rq.job import Job

job = Job.create(count_words_at_url, 'http://nvie.com')
print('Job id: %s' % job.id)
q.enqueue_job(job)

# create a job with a predetermined id
job = Job.create(count_words_at url, 'http://nvie.com', id='my_job_id')
```

This way of doing it fits my needs _very_ good, but if I do this job dependencies does not work. If this is by design then of course I will have to use `q.enqueue( ... )`.

**Comment by thomasmatecki** [comment]
I think this issue should be reopened as it appears to still occur(new failing test):
https://github.com/rq/rq/blob/9bef2aa5a49d894bd03bd772b670e207f8edd0ef/tests/test_queue.py#L530

I agree with @jtfidje that this is not related multiple dependencies, though a few folks documented it in #1170.

The crux of the issue is that [`enqueue_job`](https://github.com/rq/rq/blob/dcbbd067f0e94637f475839eca0ae805d0b2a323/rq/queue.py#L458) does not check dependencies at all.

**Comment by LukasBommes** [comment]
In encountered the issue as well and would love to see a fix.

**Comment by GitTiago** [comment]
It seems this issue still stands
Job creation takes the depends_on parameter but enqueue_job does nothing with it.
This is in direct contradiction with the docs and the test @thomasmatecki mentioned

**Comment by selwin** [comment]
`queue.enqueue_job()` just enqueues the job without doing any dependencies check, it's designed for that. Is there a reason why you don't use `queue.enqueue()` instead?

**Comment by thomasmatecki** [comment]
I think the behavior of `enqueue_job` is at issue here. It is counterintuitive that `enqueue_job` would not respect a job's dependencies while `enqueue` and `enqueue_call` do. We could document this, but it does seem like an inconsistent API (allowing a job to be created with dependencies, then ignoring them in `enqueue_job`).

What do folks think of changing the behavior of `enqueue_job` to be consistent with `enqueue`? Keep in mind this is potentially a breaking change for some users.

**Comment by caffeinatedMike** [comment]
I fully agree and support with aligning `enqueue_job` to be consistent with `enqueue` and `enqueue_call`. This is also preventing the use of `enqueue_many` due to its reliance on `enqueue_job` in the list-generator for `jobs`.

My reason for constructing the `Job` instances beforehand is so that my tasks (which I submit in successive batches) are queued as close together as possible. I also prefer `Job.create` due to its object-oriented approach. I can create my batch of jobs and sit on them for a sec while I do few other pre-flight actions, then submit them (currently in rapid succession via `enqueue_job`). It wasn't until I stumbled upon this thread that I realized my dependencies were not behaving in an orderly fashion. So, I stuck waiting for a fix (hopefully this can be address in this issue).

**Comment by caffeinatedMike** [comment]
Wouldn't moving the following snippet from `enqueue_call` into a standalone function and applying it to all the above methods provide consistency across all the aforementioned methods of the API and resolve this issue?
```python3
    job = self.setup_dependencies(
            job,
            pipeline=pipeline
        )
        # If we do not depend on an unfinished job, enqueue the job.
        if job.get_status(refresh=False) != JobStatus.DEFERRED:
            return self.enqueue_job(job, pipeline=pipeline, at_front=at_front)
        return job
```

**Comment by selwin** [comment]
The intended design for `enqueue_job()` was to just push a job into a queue. I do understand the confusion though, in hindsight we should have made this into an internal API `_enqueue_job()`.

I'm ok with refactoring `enqueue_job()` to support all the functionalities supported by `enqueue()` provided that:
1. The APIs remain backward compatible
2. All the enqueue logic are centralized in one place to keep things tidy.

I'm open to accepting a PR for this. In the meantime, I can update the docs to clarify what `enqueue_job()` does and does not do.

**Comment by jtfidje** [comment]
> @selwin I fully agree and support @thomasmatecki's point with aligning `enqueue_job` to be consistent with `enqueue` and `enqueue_call`. This is also preventing the use of `enqueue_many` due to its reliance on `enqueue_job` in the list-generator for `jobs`.
> 
> My reason for constructing the `Job` instances beforehand is so that my tasks (which I submit in successive batches) are queued as close together as possible. I also prefer `Job.create` due to its object-oriented approach. I can create my batch of jobs and sit on them for a sec while I do a few other pre-flight actions, then submit them (currently in rapid succession via `enqueue_job`). It wasn't until I stumbled upon this thread that I realized my dependencies were not behaving in an orderly fashion. So, I'm stuck waiting for a fix (hopefully this can be address in this issue).

I "fixed" this by creating a custom class that inherits from the RQ one and then just patch the function with the dependency code. I've done this for a couple of things so I run my own custom Worker, Queue and Job classes 🙂

**Comment by caffeinatedMike** [comment]
@jtfidje Would you consider contributing the "fix" via a PR? I'm sure it would be greatly appreciated by all those affected by this issue, myself included.

**Comment by selwin** [comment]
What errors did you run into?

**Comment by jtfidje** [comment]
> What errors did you run into?

I followed the docs and did this:

```
vagrant init ubuntu/trusty64
vagrant up
vagrant ssh -- "sudo apt-get -y install redis-server python-dev python-pip"
vagrant ssh -- "sudo pip install --no-input redis hiredis mock"
vagrant ssh -- "(cd /vagrant; ./run_tests)"
```

Now I guess that `run_tests` has to be copied in from somewhere, but if I run `vagrant ssh -- "(cd /vagrant; ./run_tests)"` it fails cause the script does not exist.

**Comment by caffeinatedMike** [comment]
@jtfidje I'm attempting a fix for this as well (here's [a link](https://github.com/caffeinatedMike/rq/commits/enqueue_job-depends_on-fix) to my branch]. I used the two tests [mentioned earlier](#issuecomment-787350631) in this thread with my changes, but the second still fails with my changes.

Would you be up to having a look? @selwin your input is welcome and appreciated as well.

`pytest test_queue.py::TestQueue::test_enqueue_job_with_dependency_executes_both`
```console
________________________________________________________________ TestQueue.test_enqueue_job_with_dependency_executes_both ________________________________________________________________ 

self = <tests.test_queue.TestQueue testMethod=test_enqueue_job_with_dependency_executes_both>

    def test_enqueue_job_with_dependency_executes_both(self):
        """Jobs are enqueued only when their dependencies are finished."""
        from redis import Redis
        q = Queue(connection=Redis.from_url("redis://"))

        job_a = Job.create(func=say_hello, connection=q.connection)
        job_a.save()
        job = q.enqueue_job(job_a)
        self.assertEqual(job.get_status(), JobStatus.QUEUED)

        job_b = Job.create(func=say_hello, connection=q.connection, depends_on=job_a)
        job_b.save()
        job = q.enqueue_job(job_b)

        from rq.worker import SimpleWorker
        w = SimpleWorker([q], connection=q.connection)
        w.work(burst=True)

        self.assertEqual(job_a.get_status(), JobStatus.FINISHED)
>       self.assertEqual(job_b.get_status(), JobStatus.FINISHED)
E       AssertionError: 'deferred' != <JobStatus.FINISHED: 'finished'>

test_queue.py:543: AssertionError
================================================================================ short test summary info ================================================================================= 
FAILED test_queue.py::TestQueue::test_enqueue_job_with_dependency_executes_both - AssertionError: 'deferred' != <JobStatus.FINISHED: 'finished'>
=================================================================================== 1 failed in 0.47s ====================================================================================
```

**Comment by caffeinatedMike** [comment]
@jtfidje I'm attempting a fix for this as well (here's [a link](https://github.com/caffeinatedMike/rq/commits/enqueue_job-depends_on-fix) to my branch). I used the two tests [mentioned earlier](#issuecomment-787350631) in this thread with my changes, but the second test still fails.

Would you be up to having a look? @selwin your input is welcome and appreciated as well.

`pytest test_queue.py::TestQueue::test_enqueue_job_with_dependency_executes_both`
```console
________________________________________________________________ TestQueue.test_enqueue_job_with_dependency_executes_both ________________________________________________________________ 

self = <tests.test_queue.TestQueue testMethod=test_enqueue_job_with_dependency_executes_both>

    def test_enqueue_job_with_dependency_executes_both(self):
        """Jobs are enqueued only when their dependencies are finished."""
        from redis import Redis
        q = Queue(connection=Redis.from_url("redis://"))

        job_a = Job.create(func=say_hello, connection=q.connection)
        job_a.save()
        job = q.enqueue_job(job_a)
        self.assertEqual(job.get_status(), JobStatus.QUEUED)

        job_b = Job.create(func=say_hello, connection=q.connection, depends_on=job_a)
        job_b.save()
        job = q.enqueue_job(job_b)

        from rq.worker import SimpleWorker
        w = SimpleWorker([q], connection=q.connection)
        w.work(burst=True)

        self.assertEqual(job_a.get_status(), JobStatus.FINISHED)
>       self.assertEqual(job_b.get_status(), JobStatus.FINISHED)
E       AssertionError: 'deferred' != <JobStatus.FINISHED: 'finished'>

test_queue.py:543: AssertionError
================================================================================ short test summary info ================================================================================= 
FAILED test_queue.py::TestQueue::test_enqueue_job_with_dependency_executes_both - AssertionError: 'deferred' != <JobStatus.FINISHED: 'finished'>
=================================================================================== 1 failed in 0.47s ====================================================================================
```

**Comment by caffeinatedMike** [comment]
@jtfidje How have your efforts faired?

**Comment by jtfidje** [comment]
Hi sorry! I haven't looked at it after I couldn't get the tests to run. Maybe you could walk me through your setup and how you run the tests? That would be of great help :)

**Comment by caffeinatedMike** [comment]
@selwin I believe (I have yet to confirm though) this might be a bigger issue than originally thought. You state

> `queue.enqueue_job()` just enqueues the job without doing any dependencies check, it's designed for that. Is there a reason why you don't use `queue.enqueue()` instead?

Which is a troubling thought when reviewing the `Scheduler`'s [logic for enqueueing **jobs](https://github.com/rq/rq/blob/master/rq/scheduler.py#L165)** once they reach their set time. Going off of what you stated previously, does this mean that jobs with dependencies won't operate respect/wait on their dependencies when scheduled?

**Comment by caffeinatedMike** [comment]
@jtfidje I opted to utilize WSL on windows to develop. It's much simpler.

**Comment by caffeinatedMike** [comment]
@selwin I can confirm there is definitely a bug with the scheduler in regards to dependencies.

Script to reproduce
```python3
from datetime import datetime, timedelta
from redis import Redis
from rq import Queue
from rq.registry import ScheduledJobRegistry

dt = datetime.now() + timedelta(minutes=1)
q = Queue(connection=Redis.from_url('redis://'))
r = ScheduledJobRegistry(queue=q, connection=q.connection)

j1 = q.enqueue_at(dt, "time.sleep", args=(30,), description="Long task before short task")
j2 = q.enqueue_at(dt, print, args=("Hello World",), description="Short task dependent on long task", depends_on=j1)

assert j1 in r
assert j2 in r
```

With 2 workers running both tasks are executed simultaneously, instead of j2 waiting for j1 to finish.
![dependency-issue-with-scheduler](https://user-images.githubusercontent.com/22151742/157967355-ab961e9e-57f0-4aba-aab3-4df9813f076d.png)

**Comment by selwin** [comment]
@caffeinatedMike yeah, it definitely looks like a bug. If a job is scheduled with a dependency, it should always go to `DeferredJobRegistry` instead of `ScheduledJobRegistry`.

You found so many bugs it's starting to make me look bad :(

**Comment by selwin** [comment]
@caffeinatedMike obviously I was joking about the last bit. Keep up the good work 😄

**Comment by caffeinatedMike** [comment]
@selwin sorry 😅

The good news for this bug is it all seems to tie back to `enqueue_job`. Solve that and scheduling (and all the above originally mentioned enqueue methods) should work again. 

The problem is this scheduling issue and the allow_failure issue are blocking the a project that's otherwise ready for production 😔

**Comment by caffeinatedMike** [comment]
@selwin Can you have a look at my current work on this issue at https://github.com/caffeinatedMike/rq/tree/enqueue_job-depends_on-fix. I'm really struggling to resolve this. The current test results are abysmal.

**Comment by jtfidje** [comment]
@caffeinatedMike I just pulled your repo and tried running the tests (just noticed the run-in-docker thing.)
There are a lot more than two failing tests? Or have you done more changes since your message to me? :)

**Comment by jtfidje** [comment]
Can this be fixed by doing this:
  1. Remove lines 408 through 415 from `queue.py`
  2. Have the function `enqueue_call `return `self.enqueue_job(job, pipeline=pipeline, at_front=at_front)`
  3. Add the following just below line 563 in `queue.py.enqueue_job`
  ```python
job = self.setup_dependencies(
    job,
    pipeline=pipeline
)
# If we dependon an unfinished job, just return.
if job.get_status(refresh=False) == JobStatus.DEFERRED:
    return job
```

**Comment by caffeinatedMike** [comment]
@jtfidje if you review the commit history I made a bunch of changes yesterday (see my comment above) and it's causing a bunch of failures because I'm trying a bunch of different tweaks to find a solution. 

If I the dependency issues aren't resolved soon I might just have to cut bait and opt for a different package.

**Comment by caffeinatedMike** [comment]
@jtfidje can you share the entire workaround? I only saw that you shared one custom class

**Comment by OlegZv** [comment]
Just wanted to throw my +1 in here. Looking to do exactly this - have a dependency of scheduled jobs and would benefit from this fix!

**Comment by caffeinatedMike** [comment]
@OlegZv - If you're pressed for time this can be implemented with relative ease using celery. I managed to switch over in a matter or days (coming from having zero experience with celery).

**Comment by spietras** [comment]
> I "fixed" this by creating a custom class that inherits from the RQ one and then just patch the function with the dependency code. I've done this for a couple of things so I run my own custom Worker, Queue and Job classes slightly_smiling_face

I guess it's something along the lines of:

```python
from rq import Queue, scheduler, worker
from rq.job import JobStatus


class PatchedQueue(Queue):
    def setup_dependencies(self, job, pipeline=None):
        if len(job._dependency_ids) == 0 and pipeline is not None and pipeline.explicit_transaction:
            return job
        return super().setup_dependencies(job, pipeline)

    def enqueue_job(self, job, pipeline=None, at_front=False):
        job = self.setup_dependencies(job, pipeline=pipeline)
        if job.get_status(refresh=False) != JobStatus.DEFERRED:
            return super().enqueue_job(job, pipeline, at_front)
        return job


class PatchedScheduler(scheduler.RQScheduler):
    scheduler.Queue = PatchedQueue


class PatchedWorker(worker.Worker):
    worker.RQScheduler = PatchedScheduler
```

This workaround works for me if you only use queues and workers from here. I know it's not pretty and I'm not 100% sure it works in all cases. You don't really need `PatchedScheduler` and `PatchedWorker` classes, but it just somehow looked better to me.

**Comment by caffeinatedMike** [comment]
@spietras Would you be open to creating a PR and testing against the existing tests?

**Comment by spietras** [comment]
I tried to add these changes to `enqueue_job` in `Queue` and I ran the tests locally and 17 of them failed, so this method certainly doesn't work in all cases. Can't really debug it much further right now.

**Comment by lorecoccia-zcad** [comment]
Hi all,
sorry for the question.. I have this issue using Queue.enqueue(....depends_on=...) .. I see the issue is still open, is there any other method I can use instead to enqueue a job that is dependant on a list of other jobs? Or this issue is on any method ?

Thanks

**Comment by lowercase00** [comment]
Specifically regarding to the scheduler like it was [mentioned here](https://github.com/rq/rq/issues/1404#issuecomment-1065533220) by @caffeinatedMike :

I'm looking at this for the very first time, so please correct me if I got something wrong.
My impression is that the issue is the priority between `depends_on` and `enqueued_at`.

When doing this:
```python
j1 = q.enqueue_at(dt, "time.sleep", args=(5,), description="Long task before short task")
j2 = q.enqueue_at(dt, print, args=("Hello World",), description="Short task dependent on long task", depends_on=j1)
```

There's somewhat of a conflict: should you run the job at the time the scheduler told the job to run `enqueued_at`, or should you enqueue the job based on the dependency tree? Currently, we seem to be respecting the scheduled time.

If we just enqueue the dependency, not actually sending it to the scheduler like so:

```python
j1 = q.enqueue_at(dt, "time.sleep", args=(5,), description="Long task before short task")
j2 = q.enqueue(print, args=("Hello World",), description="Short task dependent on long task", depends_on=j1)
```

Now this works as expected: job2 will run after job1.
So we there's not really a way to make this consistent: you either chose to respect the `dt`, or you can chose to respect the dependency.

**Comment by lowercase00** [comment]
@caffeinatedMike sure I understand your point and do think there's room for improvement, but even in that scenario, say you have a 1hour job (extrapolating to make a point), you would lose the schedule. That's what I meant by saying that there's no way to keep both guarantees.

**Comment by caffeinatedMike** [comment]
@lowercase00 you definitely raise a good point! I think we could likely model a dependency policy model inline with how `APScheduler` handles your scenario (see [Missed job executions and coalescing](https://apscheduler.readthedocs.io/en/3.x/userguide.html#missed-job-executions-and-coalescing)).

**Comment by lowercase00** [comment]
That's a very interesting approach indeed. It does seem more complex (a brief overview of the docs, by experience with APScheduler is limited) than just have the dependency tree be the `priority` and making it so that the `priority` is an argument in which one could define whether time was of the essence vs the dependency tree.

For now, perhaps just changing this behavior + better docs explaining the potential conflict would be enough.

**Comment by lowercase00** [comment]
Thinking about this a little more, It seems what's making things confusing is the fact that the queues takes a lot of responsibility on the jobs setup (dependencies).

What if we had `setup_dependencies` to be a method of the job - it actually only changes the `job` state, it has little to do with the queue. And then when using the queue to enqueue (anything, be it a job with `enqueue_job` or a func with `enqueue_call`) the job would carry a state `is_dependencies_setup` and the queue would check that and know whether it still needs to setup dependencies or not.

That way the queue API would remain stable, it would be easier to make a consistent behavior, and one may argue it would make more sense, since this method is just changing the job state, it does nothing to the queue.

@selwin  thoughts?

**Comment by selwin** [comment]
Sorry I just saw this message.

I'm ok with moving some dependencies logic to `Job` class if this has nothing or little to do with queue logic. We can even create a separate `dependencies.py` to further separate complex logic into different chunks so it's easier to maintain.

The issue with moving this logic to the `Job` class is that `job.py` is ~1.5k lines vs `queue.py` is ~1.3k lines so it's not like `job.py` has nothing to worry about ;)
